lib/bpf/bpf.h

```diff
@@ -0,0 +1,20 @@
+#ifndef LIB_BPF_BPF_H_
+#define LIB_BPF_BPF_H_
+
+#include <sys/syscall.h>
+
+#include "lib/error/status_or.h"
+#include "lib/posix/syscall.h"
+
+namespace bpf {
+
+// bpf() syscall wrapper. No C standard library support currently exists.
+//
+// See "man 2 bpf" for details.
+inline error::StatusOr<int> Bpf(const int cmd, bpf_attr* const attr) {
```

Comment:
Should we provide a more specific type for cmd here? The return types are somewhat cleaner already, although I guess we're still exposing int instead of something more specific inside StatusOr. Or, do we wrap these calls in cleaner high level APIs later?

Reply:
To answer my own question a bit, I guess Bpf() is multiplexed, which would force a bunch of logic into the wrapper if you wanted to be specific on the return type. That said, perhaps Bpf() should be more hidden?

Reply:
You make a great point, and I'm trying to establish a standard of sorts. Here's what I'm proposing for now:

* syscalls (both with and without C standard library wrappers) must be wrapped such that they return error::Status or error::StatusOr. 
* The wrapper should be spelled the same but with Google C++ style capitalization and all underscores removed. When a prefix of the name is implied by the containing namespace it should be omitted (for example BPF_*).
* Inputs should remain unchanged with a few exceptions:
    - use std::string_view for input that is interpreted as a const array of bytes.
    - use std::string for anything that the syscall interprets as a null terminated string (std::string::c_str() makes this easy but std::string_view has no such equivalent).
    - File descriptors must be wrapped. The goal is to call GetValue() on a wrapped type as close to the point of consumption as possible.
    - output arrays of bytes must be wrapped in a std::span equivalent.

I consider any function that takes a pointer and a length/size to be an antipattern since it is so easy to make a mistake. This is the motivation for most of the above rules that alter function signatures.

Additional wrappers can then be provided to make the syscalls more useful and even better typed. The general idea is to make the lowest level syscall wrappers look like the original and do minimal work. As more layers are added eventually a user should have no idea what syscall is being invoked.

With respect to 'int cmd' and and 'uint32_t map_type', these represent a difficult situation. If we make an enum then it must be kept up to date (automation can help but there can still be version conflicts), and it will confuse users who are familiar with the existing constant names. It makes googling for the meaning of a constant harder because of the added indirection. It is also possible to wrap cmd and map_type in base::OpaqueValue, but this also complicates the interface without providing much value (the values are wrapped just before use so there is little chance of misuse).

The rule of thumb I'm using is that if a type is copied directly from the output of a syscall to the input of another syscall without interpretation then it should be wrapped (see FileDescriptor and IfIndex). However a constant that is used as the input to a single function is worth keeping in sync with the system/kernel headers.

Reply:
Perhaps the choice of namespace should make it more clear that these functions are not intended to higher level users? Namespace bpf in a project where BPF is a top level concept might have people looking here when they should should instead be looking for higher level APIs.

Reply:
I've been playing with a ::bpf::raw and ::posix::raw namespace to use for first level syscall wrappers. Any thoughts?

Reply:
I like.

Reply:
Or the other way around? raw::bpf and raw::posix?

Reply:
Another idea, sys::wapper::posix and sys::wrapper::bpf?

Reply:
I'd love to use the sys namespace, but that is already part of the standard library and I'd rather not pollute it.

I do like changing the order to put ::raw first. This also makes the standard for stripping namespace prefixes off of function names more sensible.